### PR TITLE
gh-117142: ctypes: Migrate global vars to module state

### PR DESCRIPTION
**Pull request by neonene**
Separated PR from #117181, keeping global state access:

* Move global variables into the global state
* Call functions with the global state attached

I think almost all diffs can be checked mechanically by C compilers except stylistic changes at 9342336b29f4f0594ac8a5fb82e5ea5d99419143.

**Unfixed global variables:**
|File / Func | Var name | Possible fix |
| :--- | :--- | :--- |
| malloc_closure.c <br> `Py_ffi_closure_alloc()` <br> `Py_ffi_closure_free()`| `free_list` | ~Move to module-state or<br>thunk **type-state**, used by: <br> `_ctypes_alloc_callback()` <br> `CThunkObject_dealloc()`~ |
| malloc_closure.c                           |`_pagesize`    | Reload DLL if necessary |
| cfield.c    <br> `_ctypes_get_fielddesc()` | `initialized` | Reload DLL if necessary |
| cfield.c    <br> `_ctypes_get_fielddesc()` | `formattable` | Reload DLL if necessary |
| callbacks.c <br> `Call_GetClassObject()`   | `context` <br> (unreachable) | Do not intern (unused str?) |
| callbacks.c <br> `Call_CanUnloadNow()`     | `context` <br> (unreachable) | Do not intern (unused str?) |

<br>

* Issue: gh-117142

### CODE REVIEW

File: Modules/_ctypes/_ctypes.c [hunk 1]
```diff
@@ -794,6 +794,10 @@ CDataType_from_buffer(PyObject *type, PyObject *args)
 PyDoc_STRVAR(from_buffer_copy_doc,
 "C.from_buffer_copy(object, offset=0) -> C instance\ncreate a C instance from a readable buffer");
 
+static inline PyObject *
+_GenericPyCData_new(ctypes_state *st,
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c]: Could you use e.g. `GenericPyCData_new_st`?

Identifiers with a leading underscore followed by a capital letter are reserved according to the C standard.
We do use `_Py` all over (which will probably bite us in the future). But let's not add `_G`.

File: Modules/_ctypes/callproc.c [hunk 2]
```diff
@@ -2022,7 +2022,7 @@ create_pointer_inst(PyObject *module, PyObject *arg)
         return NULL;
     }
     if (typ == NULL) {
-        typ = create_pointer_type(NULL, (PyObject *)Py_TYPE(arg));
```
Comment (encukou) [inline Modules/_ctypes/callproc.c]: This change is not needed here; so I'll comment as if it *was* needed. Feel free to leave it to future commit (but let's not forget it):

Could you also add an `assert(module)` to `create_pointer_type`?
(And in general, to any function that starts needing a previously unused argument?)

File: Modules/_ctypes/_ctypes.c [hunk 3]
```diff
@@ -1978,12 +1978,13 @@ static PyObject *CreateSwappedType(ctypes_state *st, PyTypeObject *type,
...
         return NULL;
 
-    if (st->swapped_suffix == NULL)
+    if (st->swapped_suffix == NULL) {
 #ifdef WORDS_BIGENDIAN
         st->swapped_suffix = PyUnicode_InternFromString("_le");
 #else
         st->swapped_suffix = PyUnicode_InternFromString("_be");
 #endif
+    }
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c]: Oh, this one is particularly good :)

**Comment by encukou** [comment]
> I think almost all diffs can be checked mechanically by C compilers except stylistic changes at [one commit].

I'd prefer putting stylistic changes in their own PR, but I appreciate that it makes things harder to get reviewed.
I don't think PEP-7 has anything against Yoda-conds.
Both of those are very minor nitpicks. Please consider them for the future but please don't “fix” them here. (I've already reviewed them here, if you touch them it's just more work for both of us.)

**Review by encukou — COMMENTED** [review]
Thanks! This is indeed mind-numbingly mechanical :)

Looks almost good to go.
If I have time later today, I'll fix up the `_G` and merge. Shout if you don't want that.

File: Modules/_ctypes/callproc.c (hunk 2 above)
Comment (neonene) [reply Modules/_ctypes/callproc.c]: Addressed in gh-117181: 28696c2e33ed1e690d2c29ed436b45603b434389

File: Modules/_ctypes/_ctypes.c (hunk 1 above)
Comment (neonene) [reply Modules/_ctypes/_ctypes.c]: I've changed to snake_cases as an example.

**Comment by encukou** [comment]
Thank you!

**Pull request merged**